WIP: Add the idea of "cohorts" for subfield validation

thockin opened the pull request:
On a given field or type, there may be multiple "cohorts" of validation.  The unnamed cohort (aka default cohort) is the first.  Any named cohorts are emitted as inline functions, so early-return can be used.
    
This allows k8s:subfield to be used to generate cohorts (named after the field's JSON name) which behave properly, For example:
    
```
        // +k8s:subfield(name)=+k8s:optional
        // +k8s:subfield(name)=+k8s:format=dns-label
        // +k8s:subfield(generateName)=+k8s:optional
        // +k8s:subfield(generateName)=+k8s:format=dns-label
        metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```
    
...will generate:
    
```
       errs = append(errs,
               func(fldPath *field.Path, obj, oldObj *metav1.ObjectMeta) (errs field.ErrorList) {
                       func() { // cohort name
                               if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.OptionalValue); len (e) != 0 {
                                       return // do not proceed
                               }
                               errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.DNSLabel)...)
                       }()
                       func() { // cohort generateName
                               if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.OptionalValue); len(e) != 0 {
                                       return // do not proceed
                               }
                               errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.DNSLabel)...)
                       }()
                       return
               }(fldPath.Child("metadata"), &obj.ObjectMeta, safe.Field(oldObj, func(oldObj *corev1.ReplicationController) *metav1.ObjectMeta { return &oldObj.ObjectMeta }))...)
```
    
This behavior is pretty particular to subfield, but could plausibly be useful for something else, I think.

thockin committed:
validation-gen: Set context.Member for subfield

thockin committed:
Add support for validation cohorts

On a given field or type, there may be multiple "cohorts" of validation.
The unnamed cohort (aka default cohort) is the first.  Any named cohorts
are emitted as inline functions, so early-return can be used.

This allows k8s:subfield to be used to generate cohorts (named after the
field's JSON name) which behave properly, For example:

```
    // +k8s:subfield(name)=+k8s:optional
    // +k8s:subfield(name)=+k8s:format=dns-label
    // +k8s:subfield(generateName)=+k8s:optional
    // +k8s:subfield(generateName)=+k8s:format=dns-label
    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```

...will generate:

```
       errs = append(errs,
               func(fldPath *field.Path, obj, oldObj *metav1.ObjectMeta) (errs field.ErrorList) {
                       func() { // cohort name
                               if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.OptionalValue); len (e) != 0 {
                                       return // do not proceed
                               }
                               errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.DNSLabel)...)
                       }()
                       func() { // cohort generateName
                               if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.OptionalValue); len(e) != 0 {
                                       return // do not proceed
                               }
                               errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.DNSLabel)...)
                       }()
                       return
               }(fldPath.Child("metadata"), &obj.ObjectMeta, safe.Field(oldObj, func(oldObj *corev1.ReplicationController) *metav1.ObjectMeta { return &oldObj.ObjectMeta }))...)
```

This behavior is pretty particular to subfield, but could plausibly be
useful for something else, I think.

jpbetz commented:
How does this related with validation already on the subfields? If `ObjectMeta.name` already has optional declared on it, is that ignored by subfield validation unless explicitly declared in the cohort validation?

I'm wondering if we should "hoist" optional validation declared directly on `ObjectMeta.name` into the cohort.

thockin commented:
I thought about how to "hoist" the inner validation.  I wasn't able to reconcile how to do that when it may or may not yet exist (e.g. ObjectMeta is not yet converted to `k8s:optional`).  It seems like the "outer" call-site needs to handle the "inner" not being converted yet.  I am not sure how many other things we will hoist besides "optional".

jpbetz commented:
> I thought about how to "hoist" the inner validation. I wasn't able to reconcile how to do that when it may or may not yet exist (e.g. ObjectMeta is not yet converted to `k8s:optional`). It seems like the "outer" call-site needs to handle the "inner" not being converted yet. I am not sure how many other things we will hoist besides "optional".

OK, SGTM.  I'll review the code.

jpbetz reviewed (commented):
Looks correct.  Just a few minor comments.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -1093,105 +1093,147 @@ func (g *genValidations) emitCallToOtherTypeFunc(c *generator.Context, node *typ
 // named "fldPath".
 func emitCallsToValidators(c *generator.Context, validations []validators.FunctionGen, sw *generator.SnippetWriter) {
-	// Helper func
-	sort := func(in []validators.FunctionGen) []validators.FunctionGen {
```
Yeah, it was sort of trivial before but isn't now.

thockin commented:
Addressed, this is still WIP because we don't want to merge the name/generateName part until we decide how to position the validation differences between them.  If this is happy for review I can look to finish that

jpbetz commented:
> Addressed, this is still WIP because we don't want to merge the name/generateName part until we decide how to position the validation differences between them. If this is happy for review I can look to finish that

Go for it.  I’m onboard with the approach and the implement fell out nicely, I thought.

thockin committed:
Add DNS subdomain format

thockin commented:
small update.  RC actually uses dns-subdomain, so I added that as a format and cleaned up some tests.

Still TODO:

1) add `+k8s:format=dns-subdomain-prefix` or maybe `+k8s:format-name-prefix=dns-subdomain` (which strips trailing - and passes to the underlying format)
2) add/verify tests for RC metadata to prove this works

thockin committed:
Eliminate public ValidateReplicationControllerName

Everyone who referenced it now uses the underlying function.  Clearer
and frees me up to change objectMeta validation without impacting anyone
else.

thockin committed:
WIP: Add an alt path to objectMeta validation

Before: Name validation functions return `[]string` and apimachinery
logic wraps those in `field.Invalid()`.

Problem: This makes it hard to incrementally add declarative validation,
which needs access to potential name and generateName errors.

After: New name validation functions return `field.ErrorList`, which can
be wrapped in trivial functions to do things like "mark as covered by
declarative" in just one place rather than for all types.

e.g.:

```
-       allErrs := ValidateObjectMeta(&controller.ObjectMeta, true, ValidateReplicationControllerName, field.NewPath("metadata"))
+       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, apimachineryvalidation.ValidateNameAsDNSSubdomain, field.NewPath("metadata"))
```

and then:

```
-       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, apimachineryvalidation.ValidateNameAsDNSSubdomain, field.NewPath("metadata"))
+       namefn := func(fldPath *field.Path, name string, prefix bool) field.ErrorList {
+               return apimachineryvalidation.ValidateNameAsDNSSubdomain(field.NewPath("metadata"), name, prefix).MarkCoveredByDeclarative()
+       }
+       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, namefn, field.NewPath("metadata"))
```

WIP:
1) The `ValidateObjectMeta2` name style is a placeholder for discussion.
2) If we are open to a slightly larger diff and a "breaking" change in
   apimachinery, we could just change the signature and fix everywhere
   that calls this.
3) Rather than add funcs like `ValidateNameAsDNSSubdomain()` we could
   make callers say something like
   `validate.GenerateName(validate.DNSSubdomain)` which would return a
   function of the new singature which applies the generate-name prefix
   logic.  We probably need this for declarative anyway.

thockin committed:
Move format= tests into dirs for each format

This makes it easier to add new formats which may need different test
cases.

thockin committed:
WIP: Validation cohorts in RC.metadata.generateName

thockin committed:
Validation cohort for RC.metadata.name

`generateName` will follow.

thockin committed:
WIP: Implement tag +k8s:format=generate-name(...)

This encodes the legacy `generateName` behavior which is weird but can
be fixed at another point.

WIP: Should it be `format=generate-name(...)` or
`formatGenerateName(...)` or something else?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -251,6 +251,12 @@ func ValidateEndpointsSpecificAnnotations(annotations map[string]string, fldPath
+// resource type. Prefix is true if the name will have a value appended to it.
+//
+// This is similar to ValidateNameFunc, except for the return type.
+type ValidateNameFunc2 = apimachineryvalidation.ValidateNameFunc2
```
@jpbetz from commit comments: 

Before: Name validation functions return `[]string` and apimachinery
logic wraps those in `field.Invalid()`.

Problem: This makes it hard to incrementally add declarative validation,
which needs access to potential name and generateName errors.

After: New name validation functions return `field.ErrorList`, which can
be wrapped in trivial functions to do things like "mark as covered by
declarative" in just one place rather than for all types.

e.g.:

```
-       allErrs := ValidateObjectMeta(&controller.ObjectMeta, true, ValidateReplicationControllerName, field.NewPath("metadata"))
+       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, apimachineryvalidation.ValidateNameAsDNSSubdomain, field.NewPath("metadata"))
```

and then:

```
-       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, apimachineryvalidation.ValidateNameAsDNSSubdomain, field.NewPath("metadata"))
+       namefn := func(fldPath *field.Path, name string, prefix bool) field.ErrorList {
+               return apimachineryvalidation.ValidateNameAsDNSSubdomain(field.NewPath("metadata"), name, prefix).MarkCoveredByDeclarative()
+       }
+       allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, namefn, field.NewPath("metadata"))
```

WIP:
1) The `ValidateObjectMeta2` name style is a placeholder for discussion.
2) If we are open to a slightly larger diff and a "breaking" change in
   apimachinery, we could just change the signature and fix everywhere
   that calls this.
3) Rather than add funcs like `ValidateNameAsDNSSubdomain()` we could
   make callers say something like
   `validate.GenerateName(validate.DNSSubdomain)` which would return a
   function of the new singature which applies the generate-name prefix
   logic.  We probably need this for declarative anyway.

jpbetz replied:
> 1. The `ValidateObjectMeta2` name style is a placeholder for discussion.

Approach looks good.  Naming is hard.  Maybe `ValidateMetadata` ? :frowning_face: 

> 2. If we are open to a slightly larger diff and a "breaking" change in
>    apimachinery, we could just change the signature and fix everywhere
>    that calls this.

I don't think it's worth it.  There is significant ecosystem dependence on apimachinery as a library, both for CRD and aggregated API server dev.  

Let's evolve this compatibly-- add new functionality and migrate our own call sites, but don't break downstream call sites.

> 3. Rather than add funcs like `ValidateNameAsDNSSubdomain()` we could
>    make callers say something like
>    `validate.GenerateName(validate.DNSSubdomain)` which would return a
>    function of the new singature which applies the generate-name prefix
>    logic.  We probably need this for declarative anyway.

+100.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/format.go:
```diff
@@ -46,54 +47,86 @@ func (formatTagValidator) ValidScopes() sets.Set[Scope] {
 	dnsLabelValidator     = types.Name{Package: libValidationPkg, Name: "DNSLabel"}
 	dnsSubdomainValidator = types.Name{Package: libValidationPkg, Name: "DNSSubdomain"}
+	ipSloppyValidator     = types.Name{Package: libValidationPkg, Name: "IPSloppy"}
+	generateNameValidator = types.Name{Package: libValidationPkg, Name: "GenerateName"}
```
@jpbetz From commit comments:

WIP: Should it be `format=generate-name(...)` or `formatGenerateName(...)` or something else?

jpbetz replied:
Actually, I've changed stance slightly.

I think I'm okay with `generate-name(...)` IF we formalize the grammatical structure we're introducing into our tag parsing.

thockin replied:
So we CAN write a parser for functional args.  That's not a problem.  I think your first instinct ("surprised to run into...") is the one I am concerned with.  If we will convert this into openapi, we should not be too surprising.

https://github.com/OAI/learn.openapis.org/issues/128 and https://github.com/kubernetes/kube-openapi/pull/384 both suggest sticking to `lower-case-with-dashes`.  That leaves me with:

* `+k8s:format=k8s-generated-name=k8s-dotted-name` (still irregular)
* `+k8s:format=k8s-dotted-name-prefix` (needs one named format for each format we use as a name)
* `+k8s:format=k8s-pfx-dotted-name` (where "pfx" is magical syntax meaning the same as my `(...)` notation)
* `+k8:formatGenerateName=k8s-dotted-name`

But here's another option: What if we don't validate `GenerateName` at all?  It is used to generate `Name`, which is already validated.  So the user might put in a bad `GenerateName` and get an error on `Name` - that's a little weird, I guess.  But we already have weirdness -- I think a user can specify both fields and we don't verify that they are related, just that they are both valid.  Validation happens long AFTER name generation.

I'm not sure I love it now that I write it out, but it's an option...

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5241,6 +5241,8 @@ type ReplicationController struct {
 	// +k8s:subfield(name)=+k8s:optional
 	// +k8s:subfield(name)=+k8s:format=dns-subdomain
+	// +k8s:subfield(generateName)=+k8s:optional
+	// +k8s:subfield(generateName)=+k8s:format=generate-name(dns-subdomain)
```
WIP: pending decision on tag style

jpbetz replied:
The composition here is looking really good to me.

I just realized earlier that we can map `subfield to JSON Schema "allOf" when translating this to OpenAPI.

(I've commented elsewhere on tag style topic).

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6276,7 +6276,10 @@ func ValidateServiceStatusUpdate(service, oldService *core.Service) field.ErrorL
+	namefn := func(fldPath *field.Path, name string, prefix bool) field.ErrorList {
+		return apimachineryvalidation.ValidateNameAsDNSSubdomain(fldPath, name, prefix).MarkCoveredByDeclarative()
+	}
+	allErrs := ValidateObjectMeta2(&controller.ObjectMeta, true, namefn, field.NewPath("metadata"))
```
WIP: pending decision on evolution

thockin committed:
Add test for sparse discriminated unions

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/strfmt.go:
```diff
@@ -65,3 +66,28 @@ func DNSSubdomain[T ~string](_ context.Context, op operation.Operation, fldPath
+	//
+	// This bug has existed since forever, but we could ratchet-in the proper
+	// behavior, which would be to mimic the name generation logic and validate
+	// the result.
```
+100

I'm a fan of carrying the cruft through the migration to declarative validation.  Then, after migrated, we can use declarative validation to ratchet-in the proper behavior.  This has the advantage of showcasing the benefits of declarative validation.

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/format_test.go:
```diff
@@ -0,0 +1,45 @@
+
+import "testing"
+
+func TestParseFormat(t *testing.T) {
```
have a preference toward consolidating tag grammar into a single parser, if at all possible.
 
 I would live with this, but I'd be slightly happier if we could somehow fit this into a system wide grammar, rather than adding tag specific grammar.

jpbetz commented:
Please migrate this PR to https://github.com/jpbetz/validation-gen  (I can auto-migrate issues but not PRs)
